qutebrowser/html/history.html

```diff
@@ -16,43 +16,165 @@
+            table.lastChild.appendChild(sessionSeparator);
 
-<h1>Browsing history <span class="date">{{curr_date.strftime("%a, %d %B %Y")}}</span></h1>
+            // Create new tbody
+            tbody = document.createElement("tbody");
+            table.appendChild(tbody);
+        }
+    }
 
-<table>
-    <tbody>
-    {% for url, title, time in history %}
-        <tr>
-            <td class="title"><a href="{{url}}">{{title}}</a></td>
-            <td class="time">{{time}}</td>
-        </tr>
-    {% endfor %}
-    </tbody>
-</table>
+    return tbody;
+}
+
+/**
+ * Given a history item, create and return <tr> for it.
+ * param {string} itemUrl - The url for this item
+ * param {string} itemTitle - The title for this item
+ * param {string} itemTime - The formatted time for this item
+ */
+var makeHistoryRow = function(itemUrl, itemTitle, itemTime) {
+    var row = document.createElement('tr');
+
+    var title = document.createElement('td');
+    title.className = "title";
+    var link = document.createElement('a');
+    link.href = itemUrl;
+    link.innerHTML = itemTitle;
+    title.appendChild(link);
+
+    var time = document.createElement('td');
+    time.className = "time";
+    time.innerHTML = itemTime;
+
+    row.appendChild(title);
+    row.appendChild(time);
+
+    return row;
+}
 
-<span class="pagination-link"><a href="qute://history/?date={{prev_date.strftime("%Y-%m-%d")}}" rel="prev">Previous</a></span>
-{% if today >= next_date %}
-<span class="pagination-link"><a href="qute://history/?date={{next_date.strftime("%Y-%m-%d")}}" rel="next">Next</a></span>
-{% endif %}
+var getJSON = function(url, callback) {
+    var xhr = new XMLHttpRequest();
+    xhr.open('GET', url, true);
+    xhr.responseType = 'json';
+    xhr.onload = function() {
+      var status = xhr.status;
+      callback(status, xhr.response);
+    };
+    xhr.send();
+};
+
+/**
+ * Load new history.
+ */
+var loadHistory = function() {
+    url = "qute://history/data";
+    if (global.lastItem !== null) {
+        startTime = parseInt(global.lastItem.time) - 1;
+        url = "qute://history/data?start_time=" + startTime.toString();
+    }
+
+    getJSON(url, function(status, history) {
+        if (history !== undefined) {
+            for (item of history) {
+                atime = new Date(parseInt(item.time)*1000);
+                var session = getSessionNode(atime);
+                var row = makeHistoryRow(item.url, item.title, atime.toLocaleTimeString());
+                session.appendChild(row)
+                global.lastItem = item;
+            }
+        }
+    });
+}
+{% endblock %}
+
+{% block content %}
+<h1>Browsing history</h1>
+<div id="hist-container"></div>
+<script>
+window.onload = function() {
+    loadHistory();
 
+    window.onscroll = function(ev) {
+        if ((window.innerHeight + window.scrollY) >= document.body.scrollHeight) {
+            loadHistory();
+        }
+    };
+};
+</script>
```

Comment:
If the user doesn't have JS turned on, this will break unfortunately... Can we fall back to at least showing something about the current day in that case?